man/notify-selfcontained-example.c

```diff
@@ -0,0 +1,185 @@
+/* SPDX-License-Identifier: MIT-0 */
+
+/* Implement the systemd notify protocol without external dependencies.
+ * Supports both readiness notification on startup and on reloading,
+ * according to the protocol defined at:
+ * https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html
+ * This protocol is guaranteed to be stable as per:
+ * https://systemd.io/PORTABILITY_AND_STABILITY/ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <unistd.h>
+
+#define _cleanup_(f) __attribute__((cleanup(f)))
+
+static volatile sig_atomic_t reloading = 0;
+static volatile sig_atomic_t terminating = 0;
+
+static void closep(int *fd) {
+  if (!fd || *fd < 0)
+    return;
+
+  close(*fd);
+  *fd = -1;
+}
+
+static int notify(const char *message) {
+  union sockaddr_union {
+    struct sockaddr sa;
+    struct sockaddr_un sun;
+  } socket_addr = {
+    .sun.sun_family = AF_UNIX,
+  };
+  size_t path_length, message_length;
+  _cleanup_(closep) int fd = -1;
+  const char *socket_path;
+
+  assert(message);
+
+  message_length = strlen(message);
+  if (message_length == 0) {
+    fprintf(stderr, "Message to send to $NOTIFY_SOCKET is empty\n");
+    return -EINVAL;
+  }
+
+  socket_path = getenv("NOTIFY_SOCKET");
+  if (!socket_path)
+    return 0; /* Not running under systemd? Nothing to do */
+
+  /* Only AF_UNIX is supported, with path or abstract sockets */
+  if (socket_path[0] != '/' && socket_path[0] != '@')
+    return -EAFNOSUPPORT;
+
+  path_length = strlen(socket_path);
+  /* Ensure there is room for NUL byte */
+  if (path_length >= sizeof(socket_addr.sun.sun_path)) {
+    fprintf(stderr, "$NOTIFY_SOCKET contains a path that is too long for AF_UNIX\n");
```

Comment:
switching from perror() to fprintf() won't help you. errno may still be touched, as per POSIX.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/fprintf.html

Reply:
If an output error was encountered, these functions shall return a negative value [[CX](javascript:open_code('CX'))] [Option Start]  and set errno to indicate the error. [Option End]

Reply:
Sure, error handling needs to go recursively all the way down and check that fprintf also doesn't fail, and that the check of the check doesn't fail, and that the check of the check of the check... Let's not go overkill on this please, it's not supposed to be something that you deploy in production without looking at. Every project does error handling differently anyway, so it's moot. We can just assume fprintf to stderr is always going to work here, and that should be good enough.

Reply:
I'd just add this:

```
static int log(int error, const char *format, ...) {
        int saved_errno = errno;
        va_list ap;
        va_start(ap, format;
        errno = error;
        vfprintf(stderr, format, ap);
        va_end(ap);
        errno = saved_errno;
        return -abs(error);
}
```

i.e. have a light-weigh reimplementation of our own log_error(), that even allows you to use %m properly

Reply:
dunno, i would not publish example code that is wrong. it gets copied everywhere, and we are to blame if people don't fix this.

if you add log() as I proposed the code gets a lot shorter, as you can shorten the fprintf+return combination into one line

Reply:
> I have dropped all logging from the internal function, and just a generic error from the main loop is written instead.

that sounds ok to me.

Reply:
> Maybe it's because it's not that simple. I'm not sure what the point of this example is if it makes much more sense to go and look at libsystemd to see how it's actually done.

Well, I think it primarily teaches us that POSIX is a terrible API, that is very hard to use correctly without a lot of local glue.

Reply:
> that is very hard to use correctly without a lot of local glue.

I agree but the premise behind this PR was to provide a copy-pastable example. Now It turns out that it isn't exactly copy-pastable and should be reviewed thoroughly because it isn't even compile-tested upstream.